File: monte_carlo_pi_par.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <omp.h>

// Function to calculate Pi using Monte Carlo method
double monte_carlo_pi(int num_samples) {
    int inside_circle = 0;

    // Seed the random number generator for each thread
    #pragma omp parallel
    {
        unsigned int seed = time(NULL) ^ omp_get_thread_num();
        int local_count = 0;

        #pragma omp for
        for (int i = 0; i < num_samples; i++) {
            // Generate random x and y between -1 and 1
            double x = (double)rand_r(&seed) / RAND_MAX * 2.0 - 1.0;
            double y = (double)rand_r(&seed) / RAND_MAX * 2.0 - 1.0;

            // Check if the point is inside the circle
            if (x * x + y * y <= 1.0) {
                local_count++;
            }
        }

        #pragma omp atomic
        inside_circle += local_count;
    }

    // Calculate Pi approximation
    return (4.0 * inside_circle) / num_samples;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        printf("Usage: %s <number_of_iterations>\n", argv[0]);
        return 1;
    }

    // Parse the number of iterations from the command line
    int num_samples = atoi(argv[1]);
    if (num_samples <= 0) {
        printf("Error: Number of iterations must be a positive integer.\n");
        return 1;
    }

    // Start timing
    double start = omp_get_wtime();

    // Perform Monte Carlo Pi estimation
    double pi_estimate = monte_carlo_pi(num_samples);

    // End timing
    double end = omp_get_wtime();
    double execution_time = end - start;

    // Print results
    printf("Estimated Pi: %.8f\n", pi_estimate);
    printf("Execution Time: %.2f seconds\n", execution_time);

    return 0;
}

File: monte_carlo_pi_par.py
import random
import time
import argparse
from concurrent.futures import ProcessPoolExecutor

def monte_carlo_pi_partial(num_samples):
    """Function to calculate the number of points inside the circle for a subset of samples."""
    inside_circle = 0
    for _ in range(num_samples):
        x = random.uniform(-1, 1)
        y = random.uniform(-1, 1)
        if x**2 + y**2 <= 1:
            inside_circle += 1
    return inside_circle

def monte_carlo_pi_parallel(total_samples, num_workers):
    """Parallel Monte Carlo Pi computation."""
    samples_per_worker = total_samples // num_workers
    remaining_samples = total_samples % num_workers  # Handle remainder

    # Distribute the work across workers
    sample_counts = [samples_per_worker] * num_workers
    if remaining_samples > 0:
        sample_counts[0] += remaining_samples  # Add extra samples to the first worker

    with ProcessPoolExecutor(max_workers=num_workers) as executor:
        results = executor.map(monte_carlo_pi_partial, sample_counts)

    # Aggregate results
    total_inside_circle = sum(results)
    pi_estimate = (total_inside_circle / total_samples) * 4
    return pi_estimate

if __name__ == "__main__":
    # Argument Parsing
    parser = argparse.ArgumentParser(description="Parallel Monte Carlo Pi Simulation")
    parser.add_argument("iterations", type=int, help="Number of iterations for the simulation")
    parser.add_argument("--workers", type=int, default=4, help="Number of parallel workers (default: 4)")
    args = parser.parse_args()

    # Run Parallel Monte Carlo Simulation
    start = time.time()
    pi_estimate = monte_carlo_pi_parallel(args.iterations, args.workers)
    end = time.time()

    print(f"Estimated Pi: {pi_estimate}")
    print(f"Execution Time: {end - start:.2f} seconds")
